zne/extrapolation/polynomial_extrapolator.py

```diff
@@ -37,60 +42,128 @@ def __init__(self, degree: int = 1):  # pylint: disable=super-init-not-called
         Raises:
             ValueError: If degree is less than one.
         """
-        if degree < 1:
-            raise ValueError("Polynomial degree must be at least 1.")
-        self._degree: int = degree
+        self._set_degree(degree)
 
     ################################################################################
     ## PROPERTIES
     ################################################################################
     @property
     def degree(self) -> int:
-        """The degree of the polynomial."""
+        """The degree of the regression polynomial."""
         return self._degree
 
+    def _set_degree(self, degree: int) -> None:
+        """Degree setter."""
+        degree = int(degree)
+        if degree < 1:
+            raise ValueError("Polynomial degree must be at least 1.")
+        self._degree: int = degree
+
     ################################################################################
     ## IMPLEMENTATION
     ################################################################################
-    @cached_property
+    @property
     def min_points(self) -> int:
         return self.degree + 1
 
-    def _fit_regression_model(
-        self, data: tuple[RegressionDatum, ...]
-    ) -> tuple[RegressionModel, Metadata]:
-        x, y, _ = zip(*data)  # pylint: disable=invalid-name
-        fit = Polynomial.fit(x, y, deg=self.degree)
-        model = lambda target: (fit(target), None)  # TODO: calculate variance of fitted
-        return model, {}
+    # pylint: disable=duplicate-code
+    def _extrapolate_zero(
+        self,
+        x_data: "ndarray[Any, dtype[float_]]",
+        y_data: "ndarray[Any, dtype[float_]]",
+        sigma_x: "ndarray[Any, dtype[float_]]",
+        sigma_y: "ndarray[Any, dtype[float_]]",
+    ) -> ReckoningResult:
+        # TODO: if curve fit fails (e.g. p-value test) warn and return closest to zero
+        regression_data = _RegressionData(x_data, y_data, sigma_x, sigma_y)
+        return self._infer(0, regression_data)
+
+    ################################################################################
+    ## AUXILIARY
+    ################################################################################
+    def _infer(self, target: float, regression_data: _RegressionData) -> ReckoningResult:
+        """Fit regression model from data and infer evaluation for target value."""
+        coefficients, covariance_matrix = curve_fit(
+            self._model,
+            regression_data.x_data,
```

Comment:
Not sure I follow, I think `collections.namedtuple`s are immutable, right?

Reply:
```
from collections import namedtuple
MyTuple = namedtuple("MyTuple", ("x_data", "y_data", "sigma_x", "sigma_y"))

a = [1, 2, 3]
b = [4, 5, 6]
c = [7, 8, 9]
d = [10, 11, 12]

my_tuple = MyTuple(a, b, c, d)
a.append(1)
print(my_tuple)
```

Your namedtuple has been mutated by calling its class fields' internal methods.

This may be fine for you; I just wanted to point it out, because sometimes you absolutely don't want the user to be able to mutate something that will cause sneaky bugs not handled well by your code.

Reply:
These all seem to be sequences of floats, so you could just make them all tuples inside the `namedtuple`, and you should have full immutability then, since floats, tuples, and namedtuples are all immutable

Reply:
Ah! I follow now, you mean what happens if the tuple is made out of references to mutable objects... well it can occur, since the inner types are numpy arrays, but for internal code I don't think we should cast them back to tuples and again to `np.array`s for operation.

I am generally in favor of using tuples as much as possible, especially when interacting with users, but for the internal pipeline it is best to retain the types that make it the most efficient, in this case `np.array`s. Notice that they are casted back to tuples once they are returned to the user.

I am not particularly fond of passing `np.array`s all over though, still thinking if there might be a better solution.

Reply:
> but for internal code I don't think we should cast them back to tuples and again to `np.array`s for operation.
> 


Totally agree, you shouldn't do any excessive casting, but if you're building the sequences, and it costs you nothing, you might choose tuples. I'm assuming from your comment, it would require casting to and from tuples

Reply:
That is correct. Totally agree on using tuples always and everywhere unless there is a very good reason not to 🙂

Reply:
Named tuple now holds tuples inside